code/modules/clothing/glasses/engine_goggles.dm

```diff
@@ -39,6 +41,14 @@
 		REMOVE_TRAIT(user, TRAIT_MESON_VISION, "meson_glasses[UID()]")
 		active_on_equip = FALSE
 
+	if(mode == MODE_RAD)
+		if(!HAS_TRAIT_FROM(user, SM_HALLUCINATION_IMMUNE, "meson_glasses[UID()]"))
+			ADD_TRAIT(user, SM_HALLUCINATION_IMMUNE, "meson_glasses[UID()]")
+		active_on_equip_rad = TRUE
+	else
+		REMOVE_TRAIT(user, SM_HALLUCINATION_IMMUNE, "meson_glasses[UID()]")
+		active_on_equip_rad = FALSE
```

Comment:
This won't actually do anything, I think you need to apply it to the equipped proc too

Reply:
It seemed to work in testing, unless I'm missing something important.

Reply:
The original active_on_equip added the effect when the object was equipped, otherwise it will only be applied when the mode is changed. Also, you're going to want to make sure the trait is removed when the glasses are taken off.

Reply:
It seems I'm a silly goose - fixed.